File: CS/SampleCore/Models/LargeDatabaseUnitOfWork.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using DevExpress.Xpo;
using DevExpress.Xpo.DB;
using DevExpress.Xpo.Metadata;
using Microsoft.Extensions.DependencyInjection;

namespace SampleCore.Models
{
    public class LargeDatabaseUnitOfWork : UnitOfWork {
        public LargeDatabaseUnitOfWork(IDataLayer dataLayer)
            : base(dataLayer) {
        }

        public XPQuery<XPEmail> Emails => this.Query<XPEmail>();
    }

    [Persistent("Emails")]
    public class XPEmail : XPLiteObject {
        public XPEmail(Session session)
            : base(session) {
        }

        [Key(true)]
        public int ID {
            get => GetPropertyValue<int>("ID");
            set => SetPropertyValue("ID", value);
        }
        [Size(100)]
        public string Subject {
            get => GetPropertyValue<string>("Subject");
            set => SetPropertyValue("Subject", value);
        }
        [Size(32)]
        public string From {
            get => GetPropertyValue<string>("From");
            set => SetPropertyValue("From", value);
        }
        public DateTime Sent {
            get => GetPropertyValue<DateTime>("Sent");
            set => SetPropertyValue("Sent", value);
        }
        public long Size {
            get => GetPropertyValue<long>("Size");
            set => SetPropertyValue("Size", value);
        }
        public bool HasAttachment {
            get => GetPropertyValue<bool>("HasAttachment");
            set => SetPropertyValue("HasAttachment", value);
        }
    }

    public static class XpoServiceExtensions {
        static readonly Type[] EntityTypes = new Type[] { typeof(XPEmail) };
        static ReflectionDictionary ReflectionDictionary = new ReflectionDictionary();

        static XpoServiceExtensions() {
            ReflectionDictionary.GetDataStoreSchema(EntityTypes);
        }

        public static IServiceCollection AddLargeDatabaseUnitOfWork(this IServiceCollection services, string connectionString) {
            services.AddSingleton<LargeDatabaseDataLayer>(serviceProvider => {
                return CreatePooledDataLayer(connectionString);
            });
            services.AddScoped<LargeDatabaseUnitOfWork>(serviceProvider => {
                var dataLayer = serviceProvider.GetService<LargeDatabaseDataLayer>();
                return new LargeDatabaseUnitOfWork(dataLayer);
            });
            return services;
        }
        static LargeDatabaseDataLayer CreatePooledDataLayer(string connectionString) {
            using(var updateDataLayer = XpoDefault.GetDataLayer(connectionString, ReflectionDictionary, AutoCreateOption.DatabaseAndSchema)) {
                updateDataLayer.UpdateSchema(false, ReflectionDictionary.CollectClassInfos(EntityTypes));
            }

            var dataStore = XpoDefault.GetConnectionProvider(XpoDefault.GetConnectionPoolString(connectionString), AutoCreateOption.SchemaAlreadyExists);
            return new LargeDatabaseDataLayer(ReflectionDictionary, dataStore);
        }
    }
    public class LargeDatabaseDataLayer : ThreadSafeDataLayer {
        public LargeDatabaseDataLayer(XPDictionary dictionary, IDataStore provider)
            : base(dictionary, provider) {
        }
    }
}

File: VB/SampleCore/Models/LargeDatabaseUnitOfWork.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Threading.Tasks
Imports DevExpress.Xpo
Imports DevExpress.Xpo.DB
Imports DevExpress.Xpo.Metadata
Imports Microsoft.Extensions.DependencyInjection

Namespace SampleCore.Models
    Public Class LargeDatabaseUnitOfWork
        Inherits UnitOfWork

        Public Sub New(ByVal dataLayer As IDataLayer)
            MyBase.New(dataLayer)
        End Sub

        Public ReadOnly Property Emails() As XPQuery(Of XPEmail)
            Get
                Return Me.Query(Of XPEmail)()
            End Get
        End Property
    End Class

    <Persistent("Emails")> _
    Public Class XPEmail
        Inherits XPLiteObject

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        <Key(True)> _
        Public Property ID() As Integer
            Get
                Return GetPropertyValue(Of Integer)("ID")
            End Get
            Set(ByVal value As Integer)
                SetPropertyValue("ID", value)
            End Set
        End Property
        <Size(100)> _
        Public Property Subject() As String
            Get
                Return GetPropertyValue(Of String)("Subject")
            End Get
            Set(ByVal value As String)
                SetPropertyValue("Subject", value)
            End Set
        End Property
        <Size(32)> _
        Public Property From() As String
            Get
                Return GetPropertyValue(Of String)("From")
            End Get
            Set(ByVal value As String)
                SetPropertyValue("From", value)
            End Set
        End Property
        Public Property Sent() As Date
            Get
                Return GetPropertyValue(Of Date)("Sent")
            End Get
            Set(ByVal value As Date)
                SetPropertyValue("Sent", value)
            End Set
        End Property
        Public Property Size() As Long
            Get
                Return GetPropertyValue(Of Long)("Size")
            End Get
            Set(ByVal value As Long)
                SetPropertyValue("Size", value)
            End Set
        End Property
        Public Property HasAttachment() As Boolean
            Get
                Return GetPropertyValue(Of Boolean)("HasAttachment")
            End Get
            Set(ByVal value As Boolean)
                SetPropertyValue("HasAttachment", value)
            End Set
        End Property
    End Class

    Public Module XpoServiceExtensions
        Private ReadOnly EntityTypes() As Type = { GetType(XPEmail) }
        Private ReflectionDictionary As New ReflectionDictionary()

        Sub New()
            ReflectionDictionary.GetDataStoreSchema(EntityTypes)
        End Sub

        <System.Runtime.CompilerServices.Extension> _
        Public Function AddLargeDatabaseUnitOfWork(ByVal services As IServiceCollection, ByVal connectionString As String) As IServiceCollection
            services.AddSingleton(Of LargeDatabaseDataLayer)(Function(serviceProvider)
                Return CreatePooledDataLayer(connectionString)
            End Function)
            services.AddScoped(Of LargeDatabaseUnitOfWork)(Function(serviceProvider)
                Dim dataLayer = serviceProvider.GetService(Of LargeDatabaseDataLayer)()
                Return New LargeDatabaseUnitOfWork(dataLayer)
            End Function)
            Return services
        End Function
        Private Function CreatePooledDataLayer(ByVal connectionString As String) As LargeDatabaseDataLayer
            Using updateDataLayer = XpoDefault.GetDataLayer(connectionString, ReflectionDictionary, AutoCreateOption.DatabaseAndSchema)
                updateDataLayer.UpdateSchema(False, ReflectionDictionary.CollectClassInfos(EntityTypes))
            End Using

            Dim dataStore = XpoDefault.GetConnectionProvider(XpoDefault.GetConnectionPoolString(connectionString), AutoCreateOption.SchemaAlreadyExists)
            Return New LargeDatabaseDataLayer(ReflectionDictionary, dataStore)
        End Function
    End Module
    Public Class LargeDatabaseDataLayer
        Inherits ThreadSafeDataLayer

        Public Sub New(ByVal dictionary As XPDictionary, ByVal provider As IDataStore)
            MyBase.New(dictionary, provider)
        End Sub
    End Class
End Namespace
